Spenceral/node transfer

spenceral opened the pull request:
I'll add some unit tests but I wanted to go ahead and throw this up to get some eyes on it.

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -228,12 +248,24 @@ func (rc *replicationController) meetDesires() error {
 	switch {
 	case rc.ReplicasDesired > len(current):
+		if rc.isWatchingHealth {
+			// Remove newNode from eligible
```
might be worth including some detail in the comment for future readers

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -228,12 +248,24 @@ func (rc *replicationController) meetDesires() error {
 		}
 		nodesChanged = true
 	case len(current) > rc.ReplicasDesired:
+		if rc.isWatchingHealth {
```
ditto

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
-	if err != nil {
-		return util.Errorf("Unable to transfer nodes: %s", err)
+
+	if rc.isWatchingHealth {
```
should this check happen earlier in this function? is there a case where this can be true but we haven't run `scheduleWithSession()` etc?

spenceral replied:
Yeah there's no use having it so late

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
 	writeCtx, writeCancel := transaction.New(context.Background())
 	defer writeCancel()
-	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), modifyIndex, status)
+	err = rc.rcStatusStore.CASTxn(writeCtx, rc.ID(), 0, status)
```
I think we shouldn't assume that the status doesn't exist because we might want to track other information in the status key someday (for example maybe we want to store the full list of current nodes to reduce the cost of looking that up)

mpuncel replied:
We could consider using a different status key in consul for that though to make coordination here easier

spenceral replied:
I think ultimately my decision to do it this way (can assume non-existence of status; always delete the status when transfer is done) was because of frustration with `CASTxn`. It seemed like a lot do a `Get`, `CASTxn`, and then a `transaction.Commit` just to try and eliminate history of the node transfer. Its particularly verbose since each of those can return `error`. I preferred just being able to delete and use 0 as the index. That is my argument in favor of having a separate status. That couldv'e been laziness and I understand that might just defer complication by requiring a separate status down the road.

If we keep this as the only rc status, should I try to identify where a `Set` would work, or could a `CAS` without a transaction ever be appropriate? Or is `CASTxn` the best route?

mpuncel replied:
as a general statement, I feel like using transactional operations even if they're the only operation is still good because it makes it easier to add more operations to the transaction later on if it makes sense.

I think the argument for keeping separate status records is pretty strong, we can accomplish this by using different namespaces when configuring the status store. We should probably make the namespace string mention node transfers

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+		return err
+	}
+
+	key := path.Join(string(consul.INTENT_TREE), newNode.String(), manifest.ID().String())
```
this string computation should probably be put in `pkg/store/consul` somewhere with all the safety checks around empty hostname and pod ID, etc

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+	}
+
+	key := path.Join(string(consul.INTENT_TREE), newNode.String(), manifest.ID().String())
+	name := fmt.Sprintf("wait-on-%s-health", newNode)
```
i'd maybe put the RC ID in here too

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+
+	txnUnlocker, err := session.LockIfKeyNotExistsTxn(writeCtx, key, manifestBytes)
+	if err != nil {
+		return nil
```
did you mean `return err` here

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,267 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+	case err != nil:
+		return err
+	case !ok:
+		return util.Errorf("could not schedule new node due to a transaction violation: %s", transaction.TxnErrorsToString(resp.Errors))
```
I think here is probably the right place to `return nil` actually and just log the error along with a message like "new RC likely scheduled this node already" or something. The reason is that this function has other cases where it will return an error so we don't want to just assume everything went okay in the step above this

spenceral replied:
if we don't return an error here the calling function won't realize that this RC failed to schedule a new node. It will keep executing and spin up the goroutine and everything

mpuncel replied:
oh I see. In that case keep it the way it is

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,277 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
-	err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+func (rc *replicationController) watchHealth() {
+	defer func() {
+		rc.isWatchingHealth = false
```
you're going to have a data race on `isWatchingHealth` because multiple goroutines write to it. You could guard it with a mutex to fix that

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,277 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+				if err != nil {
+					rc.logger.WithError(err).Errorln("Could not complete transfer transaction")
+				}
+				return
```
you should also close `healthQuitCh` and `errCh` here. I'd recommend just doing that in a `defer` right after you create the channel so you don't have to keep track of all of the places that have a `return`

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,277 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+	}
+}
+
+func (rc *replicationController) finishTransfer() error {
```
just wanted to say I like the ordering of your functions, reading this PR has been pretty easy

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,277 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+		quit:    make(chan struct{}),
+	}
+
+	go rc.watchHealth()
```
`watchHealth()` does some stuff and then calls `finishTransfer()` It might be clearer if you called something like `rc.doNodeTransfer()` here which is just something like 

```
func doNodeTransfer() {
  watchHealth()
  finishTransfer()
}
``` (obviously that's pseudocode)

That I think more clearly shows the intent here and also it means you don't have to call `finishTransfer` in multiple places from `watchHealth`

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +661,277 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+		err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+		if err != nil {
+			rc.logger.WithError(err).Errorln("Unable to send alert")
+		}
```
do we need to delete the node transfer here? we probably do otherwise the new node will be ignored potentially forever right?

spenceral replied:
It gets deleted by `quitTransfer` which is called by a defer in `watchHealth`, which is the function that calls this `finishTransfer` here. Your question is just further evidence in support of your previous comment. Something should be done about the organization of these functions to make this more clear and explicit

spenceral commented:
Added unit tests and addressed previous comments. This is now ready for review

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -96,6 +99,16 @@ type consulStore interface {
+	newNode       types.NodeName
+	oldNode       types.NodeName
+	quit          chan struct{}
+	sessionCancel func()
```
nit: use `context.CancelFunc` (which is an alias for `func()` but makes it clearer what it's for)

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -625,81 +663,280 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+	errCh := make(chan error)
+	healthQuitCh := make(chan struct{})
+	defer func() {
+		close(resultCh)
```
unfortunately i think closing this isn't safe, because the way `WatchNodeService` is written it might try to send on this channel after it is closed.

I hate to give you more work for this PR, but probably `WatchNodeService()` should create `resultCh` and `errCh` itself, return them, and then take care of closing them after it sees `healthQuitCh` close.

Then you would call it like this:

```
defer close(healthQuitCh)
resultCh, errCh := rc.healthChecker.WatchNodeService(healthQuitCh)
```
which would guarantee that we have a clean shutdown.

spenceral replied:
I'll open up a new PR refactoring `WatchNodeService` to use `WatchSingle` and to avoid the panic case

mpuncel commented on pkg/rc/replication_controller_test.go:
```diff
@@ -929,3 +933,191 @@ func TestRCDoesNotFixMembership(t *testing.T) {
 	}
 }
+
+func TestTransferRolledBackByQuitCh(t *testing.T) {
```
These are great tests, they do a good job of documenting the behavior we expect as well

mpuncel commented:
Looks great! I think just that `WatchNodeService()` thing needs to be fixed. Unfortunately that one is a blocker because it will cause panics (although the RC farm catches them with a `recover()` so it wouldn't be the end of the world).

spenceral commented on pkg/rc/replication_controller.go:
```diff
@@ -242,9 +284,18 @@ func (rc *replicationController) meetDesires() error {
+	}
+
+	// transferNodes() may have changed set of current pods so we must refresh
+	current, err = rc.CurrentPods()
```
Found a race condition here that I could never reproduce locally but happened in the CI. The node transfer finished before `rc.ensureConsistency` got called so the ineligible node was scheduled again since it was still in `current`. Not sure if that would happen in reality but it happened in the test.

mpuncel replied:
ah that's interesting. It's definitely possible for this to happen in practice (although it probably wouldn't be on the same iteration of the main loop as the node transfer was started).

We should probably make sure that `ensureConsistency()` ignores the old node of a node transfer

spenceral replied:
Yeah actually this doesn't solve the problem. The old node could still be unscheduled between when ensureConsistency() is called finishes. I'll fix this

mpuncel replied:
I also think it's okay if we don't make another `CurrentPods()` call here at all. We're never going to be able to call it at exactly the right time to detect changes, so I think we just need to make sure that we make safe choices with the initial `CurrentPods()` call at the beginning of the loop, and plan on the next iteration dealing with changes appropriately.

I think as long as we exclude the new node from `eligible` and the old node from `current` as it pertains to `ensureConsistency()` then we're good

spenceral commented:
Fixed the unsafe channel close with changes from https://github.com/square/p2/pull/950 so this is ready for review again

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +669,276 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+	defer close(healthQuitCh)
+
+	// these channels are closed by WatchPodOnNode
+	resultCh, errCh := rc.healthChecker.WatchPodOnNode(rc.nodeTransfer.newNode, podID, healthQuitCh)
```
thanks for making this change, I think it reads much better this way as well

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +669,276 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+		}
+	}
+
+	rc.isWatchingHealth = false
```
I think these lines also are going to cause a race with the places where you do
```
if rc.IsWatchingHealth {
  close(rc.nodeTransfer.quit)
}
```
Probably there should be a mutex that guards the `isWatchingHealth` and `nodeTransfer` variables so that these operations are atomic. You might also be able to get away without using `isWatchingHealth` and just checking `nodeTransfer.quit != nil` or something like that.

spenceral replied:
good call

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +669,276 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
-	rc.mu.Lock()
-	newNodes, err := rc.scheduler.AllocateNodes(rc.Manifest, rc.NodeSelector, nodesRequested)
-	rc.mu.Unlock()
+	newNodes, err := rc.scheduler.AllocateNodes(man, sel, nodesRequested)
```
I'm wondering if this spot should check if there already is a newly allocated node. I'm not sure how that would happen, I guess if the allocator server makes the allocation on a previous iteration but we end up with an error.

You could feed the diff of current and eligible down to here to see if there is already an unused allocation to use for the transfer rather than calling indiscriminately.

I'm not sure how important it is to check for this, I think it might mess up the rolling update logic if there's an extra allocation lying around unused

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -633,81 +669,276 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+			}
+		case <-rc.nodeTransfer.quit:
+			return false
+		case <-time.After(5 * time.Minute):
```
I'd log something here so we know we hit this timeout

mpuncel commented:
This looks good once the `ensureConsistency()` and `isWatchingHealth` races are fixed

mpuncel commented on pkg/rc/replication_controller.go:
```diff
@@ -228,12 +244,28 @@ func (rc *replicationController) meetDesires() error {
 		}
 		nodesChanged = true
 	case len(current) > rc.ReplicasDesired:
+		if rc.nodeTransfer.quit != nil {
```
I still think you need to do something like

```
rc.nodeTransfer.mu.Lock()
if rc.nodeTransfer.quit != nil {
    close(rc.nodeTransfer.quit)
}
rc.nodeTransfer.mu.Unlock()
```

and then below acquire the lock before doing `rc.nodeTransfer = nodeTransfer{}` which nils out the channel. Otherwise you might panic closing a nil channel

spenceral replied:
Do you think I need a new mutex specifically for nodeTransfers or can I just use the `rc.mu` that the rc struct already has

mpuncel replied:
you could use the same one, yeah. They're protecting different data but there's not a lot of contention on that one so it's fine

spenceral commented:
I think I fixed the races (completely removed `isWatchingHealth` in favor of `nodeTransfer.quit != nil`) and changed transfer behavior to use an existing allocation if possible. There is an associated test for that behavior.

spenceral commented on pkg/rc/replication_controller.go:
```diff
@@ -427,8 +474,14 @@ func (rc *replicationController) removePods(current types.PodLocations, eligible
+	// if a node transfer is in progress, the old node could be in current
+	// but could be unscheduled during this function's execution. We
+	// ignore it to prevent rescheduling
+	oldNode := rc.nodeTransfer.oldNode
```
My previous change still allowed a race because a completed node transfer would make `rc.nodeTransfer.oldNode` nil. Then if the ineligible node was still in `current` it wouldn't be detected. I believe this fixes it
